core/lib/vm_utils/src/vm_env.rs

```diff
@@ -0,0 +1,243 @@
+    pub system_env: SystemEnv,
+}
+
+/// The builder for the `VmEnv`. It allows to load the required data from the database.
+#[derive(Clone, Debug)]
+pub struct VmEnvBuilder {
+    miniblock_number: Option<MiniblockNumber>,
+    l1_batch_number: L1BatchNumber,
+    prev_batch_hash: Option<H256>,
+    fee_account: Option<Address>,
+    validation_computational_gas_limit: u32,
+    chain_id: L2ChainId,
+}
+
+impl VmEnvBuilder {
+    pub fn new(
+        l1_batch_number: L1BatchNumber,
+        validation_computational_gas_limit: u32,
+        chain_id: L2ChainId,
+    ) -> Self {
+        Self {
+            l1_batch_number,
+            validation_computational_gas_limit,
+            chain_id,
+            miniblock_number: None,
+            prev_batch_hash: None,
+            fee_account: None,
+        }
+    }
+    pub fn override_miniblock_number(mut self, miniblock_number: MiniblockNumber) -> Self {
+        self.miniblock_number = Some(miniblock_number);
+        self
+    }
+
+    pub fn override_fee_account(mut self, fee_account: Address) -> Self {
+        self.fee_account = Some(fee_account);
+        self
+    }
+
+    /// Builds the `VmEnv` for the executed L1 batch. It verifies that the previous L1 batch is sealed and miniblock
+    /// is included in the L1 batch.
+    pub async fn build_for_executed_batch(
+        mut self,
+        connection: &mut StorageProcessor<'_>,
+    ) -> anyhow::Result<VmEnv> {
+        let batch_hash = connection
+            .blocks_dal()
+            .get_l1_batch_state_root_and_timestamp(self.l1_batch_number - 1)
+            .await?
+            .context(format!(
+                "Previous L1 batch not found {}",
+                self.l1_batch_number - 1
+            ))?
+            .0;
+
+        let (
+            first_miniblock_number_included_in_l1_batch,
+            last_miniblock_number_included_in_l1_batch,
+        ) = connection
+            .blocks_dal()
+            .get_miniblock_range_of_l1_batch(self.l1_batch_number)
+            .await?
+            .context(format!("L1 batch {} not found", self.l1_batch_number))?;
+
+        self.miniblock_number = if let Some(pending_miniblock_number) = self.miniblock_number {
+            if first_miniblock_number_included_in_l1_batch >= pending_miniblock_number
+                || pending_miniblock_number >= last_miniblock_number_included_in_l1_batch
+            {
+                anyhow::bail!(
+                    "Miniblock {} is not included in L1 batch {}",
+                    pending_miniblock_number,
+                    self.l1_batch_number - 1
+                )
+            }
+            Some(pending_miniblock_number)
+        } else {
+            Some(first_miniblock_number_included_in_l1_batch)
+        };
+        self.prev_batch_hash = Some(batch_hash);
+        self.build(connection).await
+    }
+
+    /// Builds the `VmEnv` for the pending L1 batch. Please note, that this method is waiting for
```

Comment:
I'm still not sure if I understand the semantics here.
This method says that it's for pending batch, but it allows you to override the miniblock number. What would the behavior be if we do so? Will we apply all the previous miniblocks stored inside, or will we just change the number (thus creating a gap here)?

Reply:
that's why we have  a protection inside